Mathlib/RingTheory/PicardGroup.lean

```diff
@@ -0,0 +1,514 @@
+  have : Function.Surjective f' := by simpa [f'] using LinearMap.lTensor_surjective _ this
+  refine ⟨.of_surjective f' this, .of_split g' f' <| LinearMap.ext fun m ↦ ?_⟩
+  simp [f', g', show f (g 1) = 1 from DFunLike.congr_fun hg 1]
+
+instance : Module.Finite R M := (finite_projective R M).1
+instance : Projective R M := (finite_projective R M).2
+example : IsReflexive R M := inferInstance
+
+section inj_surj_bij
+
+variable {R N P}
+
+theorem lTensor_injective_iff {f : N →ₗ[R] P} :
+    Function.Injective (f.lTensor M) ↔ Function.Injective f := by
+  refine ⟨fun h ↦ ?_, Flat.lTensor_preserves_injective_linearMap _⟩
+  rw [← leftCancelEquiv_comp_lTensor_comp_symm (linearEquiv R M) f]
+  simpa using Flat.lTensor_preserves_injective_linearMap _ h
+
+theorem rTensor_injective_iff {f : N →ₗ[R] P} :
+    Function.Injective (f.rTensor M) ↔ Function.Injective f := by
+  rw [← LinearMap.lTensor_inj_iff_rTensor_inj, lTensor_injective_iff]
+
+theorem lTensor_surjective_iff {f : N →ₗ[R] P} :
+    Function.Surjective (f.lTensor M) ↔ Function.Surjective f := by
+  refine ⟨fun h ↦ ?_, LinearMap.lTensor_surjective _⟩
+  rw [← leftCancelEquiv_comp_lTensor_comp_symm (linearEquiv R M) f]
+  simpa using LinearMap.lTensor_surjective _ h
+
+theorem rTensor_surjective_iff {f : N →ₗ[R] P} :
+    Function.Surjective (f.rTensor M) ↔ Function.Surjective f := by
+  rw [← LinearMap.lTensor_surj_iff_rTensor_surj, lTensor_surjective_iff]
+
+theorem lTensor_bijective_iff {f : N →ₗ[R] P} :
+    Function.Bijective (f.lTensor M) ↔ Function.Bijective f := by
+  simp_rw [Function.Bijective, lTensor_injective_iff, lTensor_surjective_iff]
+
+theorem rTensor_bijective_iff {f : N →ₗ[R] P} :
+    Function.Bijective (f.rTensor M) ↔ Function.Bijective f := by
+  simp_rw [Function.Bijective, rTensor_injective_iff, rTensor_surjective_iff]
+
+end inj_surj_bij
+
+open Finsupp in
+variable {R M} in
+/-- An invertible module is free iff it is isomorphic to the ring, i.e. its class is trivial in
+the Picard group. -/
+theorem free_iff_linearEquiv : Free R M ↔ Nonempty (M ≃ₗ[R] R) := by
+  refine ⟨fun _ ↦ ?_, fun ⟨e⟩ ↦ .of_equiv e.symm⟩
+  nontriviality R
+  have := card_eq_of_linearEquiv R <|
+    (finsuppTensorFinsupp' .. ≪≫ₗ linearEquivFunOnFinite R R _).symm ≪≫ₗ TensorProduct.congr
+      (linearEquivFunOnFinite R R _ ≪≫ₗ llift R R R _ ≪≫ₗ (Free.repr R M).dualMap)
+      (Free.repr R M).symm ≪≫ₗ linearEquiv R M ≪≫ₗ (.symm <| .funUnique Unit R R)
+  have : Unique (Free.ChooseBasisIndex R M) :=
+    (Fintype.card_eq_one_iff_nonempty_unique.mp (by simpa using this)).some
+  exact ⟨Free.repr R M ≪≫ₗ LinearEquiv.finsuppUnique R R _⟩
+
+/- TODO: generalize ≤ to arbitrary invertible modules over any commutative semiring by considering
+the localization at a prime, once we show inverible modules over local semirings are free.
+It is not clear whether ≥ holds in general but it holds for domains. -/
+theorem finrank_eq_one [StrongRankCondition R] [Free R M] : finrank R M = 1 := by
```

Comment:
If we define the rank of a module to be the maximum or minimum of local ranks, then every invertible module (not necessarily free) has rank 1. But in mathlib ranks are defined in terms of size of linearly independent sets, and I'm not sure whether one of the directions holds in full generality (my AIs fail to give a counterexample so far).

Reply:
The TODO comment isn't completely substantiated as I don't know whether local comm. semirings satisfy the strong rank condition, so the wording has been changed. If there's no further comments I'll merge in one day.